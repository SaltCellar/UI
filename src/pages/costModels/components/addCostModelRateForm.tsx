import {
  FormGroup,
  FormSelect,
  FormSelectOption,
  InputGroup,
  InputGroupText,
  TextInput,
} from '@patternfly/react-core';
import { DollarSignIcon } from '@patternfly/react-icons';
import { MetricHash } from 'api/metrics';
import { Option } from 'pages/costModels/components/logic/types';
import React from 'react';
import { InjectedTranslateProps } from 'react-i18next';

export const isRateValid = (rate: string) =>
  (!isNaN(Number(rate)) && Number(rate) > 0) || rate === '';
export const canSubmit = (rate: string) =>
  rate === '' || isNaN(Number(rate)) || Number(rate) <= 0;

export const unusedRates = (
  metricsHash: MetricHash,
  tiers: { metric: string; measurement: string }[]
) => {
  const tiersMap = tiers.reduce((acc, curr) => {
    const measureObj =
      acc[curr.metric] === undefined
        ? { [curr.measurement]: true }
        : { ...acc[curr.metric], [curr.measurement]: true };
    return {
      ...acc,
      [curr.metric]: measureObj,
    };
  }, {});
  return Object.keys(metricsHash).reduce((acc, mtr) => {
    const availableMeasurements = Object.keys(metricsHash[mtr])
      .filter(
        msr => tiersMap[mtr] === undefined || tiersMap[mtr][msr] === undefined
      )
      .map(msr => metricsHash[mtr][msr]);
    if (availableMeasurements.length === 0) {
      return acc;
    }
    return {
      ...acc,
      [mtr]: availableMeasurements.reduce((acc_, curr_) => {
        return { ...acc_, [curr_.label_measurement]: true };
      }, {}),
    };
  }, {});
};

interface CategorySelectorProps {
  isInvalid?: boolean;
  label: React.ReactNode;
  id: string;
  value: string;
  onChange: (value: string) => void;
  defaultOption?: Option;
  options: Option[];
  isDisabled?: boolean;
  testId?: string;
  helperText?: React.ReactNode;
}

const CategorySelector: React.SFC<CategorySelectorProps> = ({
  label,
  id,
  isDisabled = false,
  value,
  onChange,
  defaultOption,
  options,
  isInvalid = false,
  testId,
  helperText,
}) => (
  <FormGroup
    data-testid={testId}
    label={label}
    fieldId={id}
    helperText={helperText}
  >
    <FormSelect
      validated={!isInvalid ? 'default' : 'error'}
      isDisabled={isDisabled}
      value={value}
      onChange={onChange}
      aria-label={`form selector ${label}`}
      id={id}
    >
      <FormSelectOption
        isDisabled
        value={defaultOption.value}
        label={defaultOption.label}
      />
      {options.map(opt => (
        <FormSelectOption
          key={`${opt.value}`}
          value={opt.value}
          label={opt.label}
        />
      ))}
    </FormSelect>
  </FormGroup>
);

export interface CostTypeSelectorBaseProps extends InjectedTranslateProps {
  value: string;
  onChange: (value: string) => void;
  costTypes: string[];
}

export const CostTypeSelectorBase: React.SFC<CostTypeSelectorBaseProps> = ({
  t,
  value,
  onChange,
  costTypes,
}) => {
  const options = costTypes.map(costType => ({
    value: costType,
    label: t(`cost_models.${costType}`),
  }));

  const defaultOption = {
    value,
    label: t(`cost_models.${value}`),
  };
  return (
    <CategorySelector
      label={t('cost_models.add_rate_form.cost_type')}
      id={'rate-cost-type-selector'}
      value={value}
      onChange={onChange}
      defaultOption={defaultOption}
      options={options}
      helperText={<a href="">{t('cost_models.learn_more')}</a>}
    />
  );
};

interface SelectorBaseProps extends InjectedTranslateProps {
  options: Option[];
  onChange: (value: string) => void;
  value: string;
  isDisabled?: boolean;
  isInvalid?: boolean;
}

export const MetricSelectorBase: React.SFC<SelectorBaseProps> = ({
  t,
  value,
  onChange,
  isDisabled = false,
  options,
  isInvalid = false,
}) => {
  return (
    <CategorySelector
      testId={'metric-selector'}
      label={t(`cost_models.add_rate_form.metric_select`)}
      id={'metric-selector'}
      value={value}
      onChange={onChange}
      defaultOption={{
        label: t('cost_models.add_rate_form.default_option'),
        value: '',
      }}
      options={options}
      isDisabled={isDisabled}
      isInvalid={isInvalid}
    />
  );
};

const MeasurementSelectorBase: React.SFC<SelectorBaseProps> = ({
  t,
  value,
  onChange,
  isDisabled = false,
  options,
  isInvalid = false,
}) => {
  return (
    <CategorySelector
      testId={'measurement-selector'}
      label={t(`cost_models.add_rate_form.measurement_select`)}
      id={'measurement-selector'}
      value={value}
      onChange={onChange}
      defaultOption={{
        label: t('cost_models.add_rate_form.default_option'),
        value: '',
      }}
      options={options}
      isDisabled={isDisabled}
      isInvalid={isInvalid}
    />
  );
};

interface InputBase extends InjectedTranslateProps {
  onChange: (value: string) => void;
  value: string;
  isInvalid?: boolean;
}

const RateInputBase: React.SFC<InputBase> = ({
  t,
  value,
  onChange,
  isInvalid = false,
}) => {
  return (
    <FormGroup
      label={t('cost_models.add_rate_form.rate_input')}
      fieldId="rate-input"
      helperTextInvalid={t('cost_models.add_rate_form.error_message')}
      validated={!isInvalid ? 'default' : 'error'}
    >
      <InputGroup>
        <InputGroupText>
          <DollarSignIcon />
        </InputGroupText>
        <TextInput
          type="text"
          aria-label={t('cost_models.add_rate_form.rate_input')}
          id="rate-input"
          placeholder="0.00"
          value={value}
          onChange={onChange}
          validated={!isInvalid ? 'default' : 'error'}
        />
      </InputGroup>
    </FormGroup>
  );
};

export interface AddCostModelRateFormProps extends InjectedTranslateProps {
  metric: string;
  setMetric: (value: string) => void;
  measurement: string;
  setMeasurement: (value: string) => void;
  rate: string;
  setRate: (value: string) => void;
  metricOptions: Option[];
  measurementOptions: Option[];
  validRate: boolean;
  enableSubmit?: boolean;
  submit?: () => void;
}

export const SetMetric = ({ t, onChange, value, options }) => {
  return (
    <MetricSelectorBase
      t={t}
      onChange={onChange}
      value={value}
      options={options}
    />
  );
};

export const SetMeasurement = ({
  metricChange,
  metric,
  metricOptions,
  measurementChange,
  measurement,
  measurementOptions,
  t,
}) => {
  return (
    <>
      <MetricSelectorBase
        t={t}
        onChange={metricChange}
        value={metric}
        options={metricOptions}
      />
      <MeasurementSelectorBase
        t={t}
        onChange={measurementChange}
        value={measurement}
        options={measurementOptions}
      />
    </>
  );
};

export const SetRate = ({
  metricChange,
  metric,
  metricOptions,
  measurementChange,
  measurement,
  measurementOptions,
  isMeasurementInvalid,
  rate,
  rateChange,
  isRateInvalid,
  costTypes,
  costType,
  costTypeChange,
  t,
}) => {
  return (
    <>
      <MetricSelectorBase
        t={t}
        onChange={metricChange}
        value={metric}
        options={metricOptions}
      />
      <MeasurementSelectorBase
        t={t}
        onChange={measurementChange}
        value={measurement}
        options={measurementOptions}
        isInvalid={isMeasurementInvalid}
      />
      <RateInputBase
        t={t}
        value={rate}
        onChange={rateChange}
        isInvalid={isRateInvalid}
      />
      <CostTypeSelectorBase
        t={t}
        costTypes={costTypes}
        value={costType}
        onChange={costTypeChange}
      />
    </>
  );
};
