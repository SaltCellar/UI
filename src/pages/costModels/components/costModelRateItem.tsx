import {
  DataListAction,
  DataListCell,
  DataListItem,
  DataListItemCells,
  DataListItemRow,
  Title,
} from '@patternfly/react-core';
import React from 'react';
import { InjectedTranslateProps, translate } from 'react-i18next';
import { formatCurrency } from 'utils/rateCurrency';

export interface CostModelRateItemProps extends InjectedTranslateProps {
  index: number;
  units: string;
  metric: string;
  measurement: string;
  rate: string;
  actionComponent?: React.ReactNode;
}

const CostModelRateItemBase: React.SFC<CostModelRateItemProps> = ({
  t,
  index,
  units,
  metric,
  measurement,
  rate,
  actionComponent,
}) => {
  const unitsLabel = t(`cost_models.${units}`);
  return (
    <DataListItem aria-labelledby={`rate-${index}`}>
      <DataListItemRow>
        <DataListItemCells
          dataListCells={[
            <DataListCell key={`rate-data`}>
              <Title headingLevel="h2" size="lg">
                {t(`cost_models.${metric}`)}{' '}
                {t(`cost_models.lowercase.${measurement}`, {
                  units: unitsLabel,
                })}
              </Title>
              <Title headingLevel="h2" size="md">
                {t(`cost_models.for_every`, {
                  units: unitsLabel,
                  rate: formatCurrency(Number(rate)),
                })}
              </Title>
            </DataListCell>,
          ]}
        />
        {Boolean(actionComponent) && (
          <DataListAction
            aria-label={`actions rate-${index}`}
            id={`rate-actions-${index}`}
            aria-labelledby={`rate-${index}`}
          >
            {actionComponent}
          </DataListAction>
        )}
      </DataListItemRow>
    </DataListItem>
  );
};

export default translate()(CostModelRateItemBase);
