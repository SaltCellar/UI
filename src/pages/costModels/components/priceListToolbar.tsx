import {
  Toolbar,
  ToolbarContent,
  ToolbarFilter,
  ToolbarGroup,
  ToolbarItem,
  ToolbarItemVariant,
} from '@patternfly/react-core';
import React from 'react';

interface PriceListToolbarProps {
  primary: React.ReactNode;
  selected: string;
  secondaries: {
    component: React.ReactNode;
    name: string;
    onRemove: (category: string, chip: string) => void;
    filters: string[];
  }[];
  onClear: () => void;
  pagination: React.ReactNode;
  button: React.ReactNode;
}

export const PriceListToolbar: React.SFC<PriceListToolbarProps> = ({
  primary,
  secondaries,
  pagination,
  button,
  onClear,
  selected,
}) => {
  return (
    <Toolbar
      clearAllFilters={onClear}
      id="price-list-toolbar"
      style={{ marginBottom: '10px', marginTop: '10px' }}
    >
      <ToolbarContent>
        <ToolbarGroup variant="filter-group">
          <ToolbarItem>{primary}</ToolbarItem>
          {secondaries.map(secondary => {
            return (
              <ToolbarItem key={secondary.name}>
                <ToolbarFilter
                  deleteChip={secondary.onRemove}
                  chips={secondary.filters}
                  categoryName={secondary.name}
                >
                  {selected === secondary.name ? secondary.component : ''}
                </ToolbarFilter>
              </ToolbarItem>
            );
          })}
        </ToolbarGroup>
        <ToolbarItem>{button}</ToolbarItem>
        <ToolbarItem variant={ToolbarItemVariant.pagination}>
          {pagination}
        </ToolbarItem>
      </ToolbarContent>
      <hr className="pf-c-divider" />
    </Toolbar>
  );
};
