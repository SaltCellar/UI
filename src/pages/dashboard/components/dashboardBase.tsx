import { Grid, GridItem } from '@patternfly/react-core';
import React from 'react';
import { WrappedComponentProps } from 'react-intl';

type DashboardOwnProps = WrappedComponentProps;

interface DashboardStateProps {
  DashboardWidget: any;
  widgets: number[];
}

interface DashboardDispatchProps {
  selectWidgets?: () => void;
}

type DashboardProps = DashboardOwnProps &
  DashboardStateProps &
  DashboardDispatchProps;

const DashboardBase: React.SFC<DashboardProps> = ({
  DashboardWidget,
  selectWidgets,
  widgets,
}) => (
  <div>
    <Grid gutter="md">
      {widgets.map(widgetId => {
        const widget = selectWidgets[widgetId];
        return Boolean(widget.details.showHorizontal) ? (
          <GridItem sm={12} key={widgetId}>
            <DashboardWidget widgetId={widgetId} />
          </GridItem>
        ) : (
          <GridItem lg={12} xl={6} xl2={4} key={widgetId}>
            <DashboardWidget widgetId={widgetId} />
          </GridItem>
        );
      })}
    </Grid>
  </div>
);

export { DashboardBase };
