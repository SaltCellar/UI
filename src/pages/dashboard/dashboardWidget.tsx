import { Report } from 'api/reports';
import {
  ReportSummary,
  ReportSummaryDetails,
  ReportSummaryTrend,
} from 'components/reportSummary';
import formatDate from 'date-fns/format';
import getDate from 'date-fns/get_date';
import getMonth from 'date-fns/get_month';
import startOfMonth from 'date-fns/start_of_month';
import React from 'react';
import { InjectedTranslateProps, translate } from 'react-i18next';
import { connect } from 'react-redux';
import { createMapStateToProps } from 'store/common';
import {
  dashboardActions,
  dashboardSelectors,
  DashboardWidget as DashboardWidgetStatic,
} from 'store/dashboard';
import { reportsSelectors } from 'store/reports';
import { formatValue } from 'utils/formatValue';

interface DashboardWidgetOwnProps {
  widgetId: number;
}

interface DashboardWidgetStateProps extends DashboardWidgetStatic {
  current: Report;
  previous: Report;
}

interface DashboardWidgetDispatchProps {
  fetchReports: typeof dashboardActions.fetchWidgetReports;
  updateTab: typeof dashboardActions.changeWidgetTab;
}

type DashboardWidgetProps = DashboardWidgetOwnProps &
  DashboardWidgetStateProps &
  DashboardWidgetDispatchProps &
  InjectedTranslateProps;

class DashboardWidgetBase extends React.Component<DashboardWidgetProps> {
  public componentDidMount() {
    const { fetchReports, widgetId } = this.props;
    fetchReports(widgetId);
  }

  public render() {
    const { t, titleKey, trend, details, current, previous } = this.props;
    const today = new Date();
    const month = getMonth(today);
    const endDate = formatDate(today, 'Do');
    const startDate = formatDate(startOfMonth(today), 'Do');

    const title = t(titleKey, { endDate, month, startDate });
    const detailLabel = t(details.labelKey, {
      context: details.labelKeyContext,
    });

    const detailDescription = t(
      getDate(today) === 1
        ? details.descriptionKeySingle
        : details.descriptionKeyRange || details.descriptionKeySingle,
      { endDate, month, startDate }
    );
    const trendTitle = t(trend.titleKey);

    return (
      <ReportSummary title={title}>
        <ReportSummaryDetails
          report={current}
          formatValue={formatValue}
          label={detailLabel}
          formatOptions={details.formatOptions}
          description={detailDescription}
        />
        <ReportSummaryTrend
          title={trendTitle}
          current={current}
          previous={previous}
          formatDatumValue={formatValue}
          formatDatumOptions={trend.formatOptions}
        />
      </ReportSummary>
    );
  }
}

const mapStateToProps = createMapStateToProps<
  DashboardWidgetOwnProps,
  DashboardWidgetStateProps
>((state, { widgetId }) => {
  const widget = dashboardSelectors.selectWidget(state, widgetId);
  const queries = dashboardSelectors.selectWidgetQueries(state, widgetId);
  return {
    ...widget,
    current: reportsSelectors.selectReport(
      state,
      widget.reportType,
      queries.current
    ),
    previous: reportsSelectors.selectReport(
      state,
      widget.reportType,
      queries.previous
    ),
  };
});

const mapDispatchToProps: DashboardWidgetDispatchProps = {
  fetchReports: dashboardActions.fetchWidgetReports,
  updateTab: dashboardActions.changeWidgetTab,
};

const DashboardWidget = translate()(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(DashboardWidgetBase)
);

export { DashboardWidget, DashboardWidgetBase, DashboardWidgetProps };
