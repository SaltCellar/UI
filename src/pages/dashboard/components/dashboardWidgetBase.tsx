import { Tab, Tabs } from '@patternfly/react-core';
import { css } from '@patternfly/react-styles';
import { getQuery } from 'api/queries/awsQuery';
import { Report } from 'api/reports/report';
import {
  ChartComparison,
  transformReport,
} from 'components/charts/common/chartUtils';
import {
  ReportSummary,
  ReportSummaryAlt,
  ReportSummaryCost,
  ReportSummaryDetails,
  ReportSummaryItem,
  ReportSummaryItems,
  ReportSummaryTrend,
  ReportSummaryUsage,
} from 'components/reports/reportSummary';
import formatDate from 'date-fns/format';
import getDate from 'date-fns/get_date';
import getMonth from 'date-fns/get_month';
import startOfMonth from 'date-fns/start_of_month';
import React from 'react';
import { InjectedTranslateProps } from 'react-i18next';
import { Link } from 'react-router-dom';
import {
  DashboardChartType,
  DashboardWidget,
} from 'store/dashboard/common/dashboardCommon';
import { formatValue, unitLookupKey } from 'utils/formatValue';
import { chartStyles, styles } from './dashboardWidget.styles';

interface DashboardWidgetOwnProps {
  appNavPath: string;
  chartAltHeight?: number;
  containerAltHeight?: number;
  detailsPath: string;
  getIdKeyForTab: <T extends DashboardWidget<any>>(tab: T) => string;
  widgetId: number;
}

interface DashboardWidgetStateProps extends DashboardWidget<any> {
  currentQuery: string;
  currentReport: Report;
  currentReportFetchStatus: number;
  previousQuery: string;
  previousReport: Report;
  tabsQuery: string;
  tabsReport: Report;
  tabsReportFetchStatus: number;
}

interface DashboardWidgetDispatchProps {
  fetchReports: (widgetId) => void;
  updateTab: (id, availableTabs) => void;
}

type DashboardWidgetProps = DashboardWidgetOwnProps &
  DashboardWidgetStateProps &
  DashboardWidgetDispatchProps &
  InjectedTranslateProps;

class DashboardWidgetBase extends React.Component<DashboardWidgetProps> {
  public state = {
    activeTabKey: 0,
  };

  public componentDidMount() {
    const { availableTabs, fetchReports, id, updateTab, widgetId } = this.props;
    if (availableTabs) {
      updateTab(id, availableTabs[0]);
    }
    fetchReports(widgetId);
  }

  private buildDetailsLink = <T extends DashboardWidget<any>>(tab: T) => {
    const { detailsPath, getIdKeyForTab } = this.props;
    const currentTab = getIdKeyForTab(tab);
    return `${detailsPath}?${getQuery({
      group_by: {
        [currentTab]: '*',
      },
      order_by: { cost: 'desc' },
    })}`;
  };

  private getChart = (
    containerHeight: number,
    height: number,
    adjustContainerHeight: boolean = false
  ) => {
    const { chartType } = this.props;
    if (chartType === DashboardChartType.cost) {
      return this.getCostChart(containerHeight, height, adjustContainerHeight);
    } else if (chartType === DashboardChartType.trend) {
      return this.getTrendChart(containerHeight, height, adjustContainerHeight);
    } else if (chartType === DashboardChartType.usage) {
      return this.getUsageChart(height);
    } else {
      return null;
    }
  };

  private getCostChart = (
    containerHeight: number,
    height: number,
    adjustContainerHeight: boolean = false
  ) => {
    const { currentReport, previousReport, t, trend } = this.props;

    const units = this.getUnits();
    const title = t(trend.titleKey, { units: t(`units.${units}`) });

    // Infrastructure data
    const currentInfrastructureData = transformReport(
      currentReport,
      trend.type,
      'date',
      'infrastructureCost'
    );
    const previousInfrastructureData = transformReport(
      previousReport,
      trend.type,
      'date',
      'infrastructureCost'
    );

    // Usage data
    const currentUsageData = transformReport(
      currentReport,
      trend.type,
      'date',
      'cost'
    );
    const previousUsageData = transformReport(
      previousReport,
      trend.type,
      'date',
      'cost'
    );

    return (
      <ReportSummaryCost
        adjustContainerHeight={adjustContainerHeight}
        containerHeight={containerHeight}
        currentCostData={currentUsageData}
        currentInfrastructureCostData={currentInfrastructureData}
        formatDatumValue={formatValue}
        formatDatumOptions={trend.formatOptions}
        height={height}
        previousCostData={previousUsageData}
        previousInfrastructureCostData={previousInfrastructureData}
        title={title}
      />
    );
  };

  private getTrendChart = (
    containerHeight: number,
    height: number,
    adjustContainerHeight: boolean = false
  ) => {
    const { currentReport, details, previousReport, t, trend } = this.props;

    const units = this.getUnits();
    const title = t(trend.titleKey, { units: t(`units.${units}`) });

    // Data
    const currentData = transformReport(
      currentReport,
      trend.type,
      'date',
      trend.comparison
    );
    const previousData = transformReport(
      previousReport,
      trend.type,
      'date',
      trend.comparison
    );

    return (
      <ReportSummaryTrend
        adjustContainerHeight={adjustContainerHeight}
        containerHeight={containerHeight}
        currentData={currentData}
        formatDatumValue={formatValue}
        formatDatumOptions={trend.formatOptions}
        height={height}
        previousData={previousData}
        showUsageLegendLabel={details.showUsageLegendLabel}
        title={title}
        units={units}
      />
    );
  };

  private getUsageChart = (height: number) => {
    const { currentReport, previousReport, t, trend } = this.props;

    const units = this.getUnits();
    const title = t(trend.titleKey, { units: t(`units.${units}`) });

    // Request data
    const currentRequestData = transformReport(
      currentReport,
      trend.type,
      'date',
      'request'
    );
    const previousRequestData = transformReport(
      previousReport,
      trend.type,
      'date',
      'request'
    );

    // Usage data
    const currentUsageData = transformReport(
      currentReport,
      trend.type,
      'date',
      'usage'
    );
    const previousUsageData = transformReport(
      previousReport,
      trend.type,
      'date',
      'usage'
    );

    return (
      <ReportSummaryUsage
        containerHeight={chartStyles.containerUsageHeight}
        currentRequestData={currentRequestData}
        currentUsageData={currentUsageData}
        formatDatumValue={formatValue}
        formatDatumOptions={trend.formatOptions}
        height={height}
        previousRequestData={previousRequestData}
        previousUsageData={previousUsageData}
        title={title}
      />
    );
  };

  private getDetails = () => {
    const {
      chartType,
      currentReport,
      details,
      isUsageFirst,
      perspective,
    } = this.props;
    const units = this.getUnits();
    return (
      <ReportSummaryDetails
        chartType={chartType}
        costLabel={this.getDetailsLabel(details.costKey, units)}
        formatOptions={details.formatOptions}
        formatValue={formatValue}
        perspective={perspective}
        report={currentReport}
        requestLabel={this.getDetailsLabel(details.requestKey, units)}
        showUnits={details.showUnits}
        showUsageFirst={isUsageFirst}
        usageFormatOptions={details.usageFormatOptions}
        usageLabel={this.getDetailsLabel(details.usageKey, units)}
      />
    );
  };

  private getDetailsLabel = (key: string, units: string) => {
    const { t } = this.props;
    return key ? t(key, { units: t(`units.${units}`) }) : undefined;
  };

  private getDetailsLink = () => {
    const { currentTab, isDetailsLink } = this.props;
    return (
      isDetailsLink && (
        <Link
          to={this.buildDetailsLink(currentTab)}
          onClick={this.handleInsightsNavClick}
        >
          {this.getDetailsLinkTitle(currentTab)}
        </Link>
      )
    );
  };

  private getDetailsLinkTitle = <T extends DashboardWidget<any>>(tab: T) => {
    const { getIdKeyForTab, t } = this.props;
    const key = getIdKeyForTab(tab) || '';

    return t('group_by.all', { groupBy: key });
  };

  private getHorizontalLayout = () => {
    const {
      containerAltHeight = chartStyles.containerAltHeight,
      chartAltHeight = chartStyles.chartAltHeight,
      currentReportFetchStatus,
    } = this.props;

    return (
      <ReportSummaryAlt
        detailsLink={this.getDetailsLink()}
        status={currentReportFetchStatus}
        subTitle={this.getSubTitle()}
        tabs={this.getTabs()}
        title={this.getTitle()}
      >
        {this.getDetails()}
        {this.getChart(containerAltHeight, chartAltHeight, true)}
      </ReportSummaryAlt>
    );
  };

  private getSubTitle = () => {
    const { t } = this.props;

    const today = new Date();
    const month = getMonth(today);
    const endDate = formatDate(today, 'D');
    const startDate = formatDate(startOfMonth(today), 'D');

    return t('aws_dashboard.widget_subtitle', {
      count: getDate(today),
      endDate,
      month,
      startDate,
    });
  };

  private getTab = <T extends DashboardWidget<any>>(tab: T, index: number) => {
    const { getIdKeyForTab, tabsReport, tabsReportFetchStatus } = this.props;
    const currentTab: any = getIdKeyForTab(tab);

    return (
      <Tab
        eventKey={index}
        key={`${getIdKeyForTab(tab)}-tab`}
        title={this.getTabTitle(tab)}
      >
        <div className={css(styles.tabItems)}>
          <ReportSummaryItems
            idKey={currentTab}
            key={`${currentTab}-items`}
            report={tabsReport}
            status={tabsReportFetchStatus}
          >
            {({ items }) =>
              items.map(reportItem => this.getTabItem(tab, reportItem))
            }
          </ReportSummaryItems>
        </div>
      </Tab>
    );
  };

  private getTabItem = <T extends DashboardWidget<any>>(tab: T, reportItem) => {
    const {
      availableTabs,
      details,
      getIdKeyForTab,
      tabsReport,
      topItems,
      trend,
    } = this.props;
    const { activeTabKey } = this.state;

    const currentTab = getIdKeyForTab(tab);
    const activeTab = getIdKeyForTab(availableTabs[activeTabKey]);

    let totalValue;
    const hasTotal = tabsReport && tabsReport.meta && tabsReport.meta.total;
    if (trend.comparison === ChartComparison.usage) {
      const hasUsage = hasTotal && tabsReport.meta.total.usage;
      totalValue = hasUsage ? tabsReport.meta.total.usage.value : undefined;
    } else {
      const hasCost = hasTotal && tabsReport.meta.total.cost;
      totalValue = hasCost ? tabsReport.meta.total.cost.total.value : undefined;
    }

    if (activeTab === currentTab) {
      return (
        <ReportSummaryItem
          key={`${reportItem.id}-item`}
          formatOptions={topItems.formatOptions}
          formatValue={formatValue}
          label={reportItem.label ? reportItem.label.toString() : ''}
          totalValue={totalValue}
          units={details.units ? details.units : reportItem.units}
          value={reportItem.cost}
        />
      );
    } else {
      return null;
    }
  };

  private getTabs = () => {
    const { availableTabs } = this.props;
    return (
      <Tabs
        isFilled
        activeKey={this.state.activeTabKey}
        onSelect={this.handleTabClick}
      >
        {availableTabs.map((tab, index) => this.getTab(tab, index))}
      </Tabs>
    );
  };

  private getTabTitle = <T extends DashboardWidget<any>>(tab: T) => {
    const { getIdKeyForTab, t } = this.props;
    const key = getIdKeyForTab(tab) || '';

    return t('group_by.top', { groupBy: key });
  };

  private getTitle = () => {
    const { t, titleKey } = this.props;

    const today = new Date();
    const month = getMonth(today);
    const endDate = formatDate(today, 'Do');
    const startDate = formatDate(startOfMonth(today), 'Do');

    return t(titleKey, { endDate, month, startDate });
  };

  private getUnits = () => {
    const { currentReport, details, trend } = this.props;

    if (details.units) {
      return details.units;
    }

    let units;
    const hasTotal =
      currentReport && currentReport.meta && currentReport.meta.total;
    if (trend.comparison === ChartComparison.usage) {
      const hasUsage = hasTotal && currentReport.meta.total.usage;
      units = hasUsage
        ? unitLookupKey(currentReport.meta.total.usage.units)
        : '';
    } else {
      const hasCost = hasTotal && currentReport.meta.total.cost;
      units = hasCost
        ? unitLookupKey(currentReport.meta.total.cost.total.units)
        : '';
    }
    return units;
  };

  private getVerticalLayout = () => {
    const { availableTabs, currentReportFetchStatus } = this.props;

    return (
      <ReportSummary
        detailsLink={this.getDetailsLink()}
        status={currentReportFetchStatus}
        subTitle={this.getSubTitle()}
        title={this.getTitle()}
      >
        {this.getDetails()}
        {this.getChart(
          chartStyles.containerTrendHeight,
          chartStyles.chartHeight
        )}
        {Boolean(availableTabs) && (
          <div className={css(styles.tabs)}>{this.getTabs()}</div>
        )}
      </ReportSummary>
    );
  };

  private handleInsightsNavClick = () => {
    const { appNavPath } = this.props;
    insights.chrome.appNavClick({ id: appNavPath, secondaryNav: true });
  };

  private handleTabClick = (event, tabIndex) => {
    const { availableTabs, id, updateTab } = this.props;
    const tab = availableTabs[tabIndex];

    updateTab(id, tab);
    this.setState({
      activeTabKey: tabIndex,
    });
  };

  public render() {
    const { isHorizontal = false } = this.props;
    return Boolean(isHorizontal)
      ? this.getHorizontalLayout()
      : this.getVerticalLayout();
  }
}

export { DashboardWidgetBase, DashboardWidgetProps };
