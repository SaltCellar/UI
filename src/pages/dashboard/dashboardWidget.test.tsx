import { ReportType } from 'api/reports';
import { ReportSummary, ReportSummaryDetails } from 'components/reportSummary';
import { shallow } from 'enzyme';
import React from 'react';
import { FetchStatus } from 'store/common';
import { DashboardWidgetBase, DashboardWidgetProps } from './dashboardWidget';

const props: DashboardWidgetProps = {
  reportType: ReportType.cost,
  fetchReport: jest.fn(),
  currentMonthFetchStatus: FetchStatus.complete,
  currentMonthError: null,
  currentMonth: { data: [] },
  currentMonthQuery: 'currentMonthQuery',
  prevMonthFetchStatus: FetchStatus.complete,
  prevMonth: { data: [] },
  prevMonthError: null,
  prevMonthQuery: 'previousMonthQuery',
  formatDetailsValue: jest.fn(),
  formatTrendValue: jest.fn(),
  title: 'Title',
  trendTitle: 'trend title',
  detailDescription: 'detail description',
  detailLabel: 'detail label',
};

test('previous and current month reports are fetched on mount', () => {
  shallow(<DashboardWidgetBase {...props} />);
  expect(props.fetchReport).toBeCalledWith(
    props.reportType,
    props.currentMonthQuery
  );
  expect(props.fetchReport).toBeCalledWith(
    props.reportType,
    props.prevMonthQuery
  );
});

test('title is passed to ReportSummary', () => {
  const view = shallow(<DashboardWidgetBase {...props} />);
  expect(view.find(ReportSummary).props().title).toBe(props.title);
});

test('current report and strings are passed to ReportSummaryDetails', () => {
  const view = shallow(<DashboardWidgetBase {...props} />);
  const detailProps = view.find(ReportSummaryDetails).props();
  expect(detailProps.report).toBe(props.currentMonth);
  expect(detailProps.formatValue).toBe(props.formatDetailsValue);
  expect(detailProps.description).toBe(props.detailDescription);
  expect(detailProps.label).toBe(props.detailLabel);
});
