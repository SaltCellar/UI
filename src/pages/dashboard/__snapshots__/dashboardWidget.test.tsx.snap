// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail description is translated with date range 1`] = `
Array [
  "detail description range",
  Object {
    "endDate": "formated date",
    "month": 1,
    "startDate": "formated date",
  },
]
`;

exports[`detail description is translated with single date 1`] = `
Array [
  "detail description single",
  Object {
    "endDate": "formated date",
    "month": 1,
    "startDate": "formated date",
  },
]
`;

exports[`detail label is translated 1`] = `
Array [
  "detail label",
  Object {
    "context": "label context",
  },
]
`;

exports[`title is translated 1`] = `
Array [
  "title",
  Object {
    "endDate": "formated date",
    "month": 1,
    "startDate": "formated date",
  },
]
`;

exports[`trend title is translated 1`] = `
Array [
  "trend title",
]
`;
